cmd/ddblibrarian-clone/main.go

```diff
@@ -0,0 +1,250 @@
+	batchSize         int = 25
+)
+
+type appConfig struct {
+	srcRegion        string
+	dstRegion        string
+	srcTable         string
+	dstTable         string
+	partitionKey     string
+	partitionKeyType string
+	rangeKey         string
+	rangeKeyType     string
+	snapshot         string
+	maxRetries       int
+}
+
+func checkFlags(app *appConfig) {
+	if app.srcTable == "" || app.dstTable == "" {
+		log.Fatal("Both source and destination tables are mandatory")
+	}
+
+	if app.partitionKey == "" {
+		log.Fatal("The partition key is required")
+	}
+
+	if app.partitionKeyType == "" {
+		log.Fatal("The partition key type (S or N) is required")
+	}
+}
+
+func connect(app *appConfig) (*dynamodb.DynamoDB, *ddblibrarian.Library) {
+	srcSession, err := session.NewSession(&aws.Config{
+		Region:     aws.String(app.srcRegion),
+		MaxRetries: aws.Int(3),
+	})
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+
+	dstSession, err := session.NewSession(&aws.Config{
+		Region:     aws.String(app.dstRegion),
+		MaxRetries: aws.Int(3),
+	})
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+
+	librarian, err := ddblibrarian.New(
+		app.dstTable,
+		app.partitionKey,
+		app.partitionKeyType,
+		app.rangeKey,
+		app.rangeKeyType,
+		dstSession,
+	)
+	if err != nil {
+		log.Fatal(err.Error())
+	}
+
+	return dynamodb.New(srcSession), librarian
+}
+
+func writeBatch(
+	batch map[string][]*dynamodb.WriteRequest,
+	library *ddblibrarian.Library,
+	maxRetries int,
+) error {
+	var err error
+
+	for i := 0; i < maxRetries; i++ {
+		_, err = library.BatchWriteItem(&dynamodb.BatchWriteItemInput{
+			RequestItems: batch,
+		})
+		if err != nil {
+			if aerr, ok := err.(awserr.Error); ok {
+				if aerr.Code() == dynamodb.ErrCodeProvisionedThroughputExceededException {
+					wait := math.Pow(2, float64(i)) * 100
+					log.Printf("BatchWriteItem: backing off for %f milliseconds\n", wait)
+					time.Sleep(time.Duration(wait) * time.Millisecond)
+					continue
+				}
+			} else {
+				// there's no point on retrying
+				return err
+			}
+		} else {
+			// the write succeeded, nothing else to do here
+			return nil
+		}
+	}
+
+	// if we've made it this far, all attempts have failed
+	return err
+}
+
+func writeItems(
+	items []map[string]*dynamodb.AttributeValue,
+	lastEvaluatedKey map[string]*dynamodb.AttributeValue,
+	library *ddblibrarian.Library,
+	app *appConfig,
+) {
+	var err error = nil
+	requests := make(map[string][]*dynamodb.WriteRequest, 0)
+
+	// create groups of 25 items -- max batch size
+	for i, item := range items {
+		if (i%batchSize) == 0 && i > 0 {
+			err = writeBatch(requests, library, app.maxRetries)
+			if err != nil {
+				break
+			}
+			requests = make(map[string][]*dynamodb.WriteRequest, 0)
+		} else {
+			requests[app.dstTable] = append(requests[app.dstTable], &dynamodb.WriteRequest{
+				PutRequest: &dynamodb.PutRequest{
+					Item: item,
+				}})
+		}
+	}
+
+	if err != nil {
+		log.Fatalln("Failed to write batch:", err)
+		// this can be quite long...
```

Comment:
Good point. Yeah this could be quite long. We can skip this for now. This might be useful in case some specific writes fail because of throttling.